Refactor product list facets

federicobadini opened the pull request:
closes #1362

ardelato commented:
The Playwright `cross-sell` test failure is unrelated to the PR. I am going to make a new PR to fix it

masonmcelvain commented:
@ardelato It looks like @federicobadini already opened a test to fix it here: https://github.com/iFixit/react-commerce/pull/1385

ardelato commented:
Ah nice! I did not see that. I'll go ahead and close my PR

masonmcelvain commented on frontend/templates/product-list/sections/FilterableProductsSection/Toolbar.tsx:
```diff
@@ -92,15 +96,27 @@ export function Toolbar(props: ToolbarProps) {
+                  align="flex-start"
+                  display={{ base: 'flex', md: 'none' }}
+               >
+                  <CurrentRefinements />
```
I noticed that the design displayed the current refinements in a single row with overflow-x hidden, whereas the implementation wrapped them onto multiple rows. It looks nice as implemented, but thought I'd point out the discrepancy.

| Spec | Implementation |
|-|-|
| ![image](https://user-images.githubusercontent.com/52104630/220207117-179720e4-c908-434f-8731-8812c5d63f10.png) | ![image](https://user-images.githubusercontent.com/52104630/220207068-5b5a31b5-91f4-4f5b-b686-689e8a28ab78.png) |

federicobadini replied:
Yep, you are right. During the development together with @mmarcon91 we discussed this.
Having chips on a single row make more room for the product list but make the current selection harder to read.
So we decided to keep the multiple row approach for a while and eventually to move to single row only if needed

masonmcelvain commented on frontend/tests/playwright/product-list/filters.spec.ts:
```diff
@@ -5,12 +5,21 @@ import { waitForAlgoliaSearch, resolvePath } from '../utils';
       throw new Error('Could not find ' + value);
    }
-   await expect(page.getByTestId(`current-refinement-${value}`)).toBeVisible();
+   if (layout === 'desktop') {
+      await expect(
+         page.getByTestId(`current-refinement-${value}`).first()
+      ).toBeVisible();
+   } else {
+      await expect(
+         page.getByTestId(`current-refinement-${value}`).last()
+      ).toBeVisible();
+   }
```
It's not clear to me why we're making the distinction between mobile and desktop in this test. When I inspect the `<CurrentRefinements>` component, it appears that all current refinements are rendered on both mobile and desktop, so I'm confused why we can't make the same assertion on both.

federicobadini replied:
We have multiple `<CurrentRefinements>` components in dom. However their visibility depends on the layout.
If we are on a mobile layout `<CurrentRefinements>` is visible only in the toolbar component.
If we are on a desktop layout `<CurrentRefinements>` is located in the upper part of the side panel.

To avoid altering the test, I made the above distinction.

masonmcelvain replied:
Thanks for the explanation! I missed the fact that there are 2 components.

masonmcelvain reviewed (approved):
CR ✌🏻 Nice work! I have a couple non-blocking questions but looks good to me

federicobadini commented:
Aligned with main to include the test fix

masonmcelvain reviewed (approved):
CR ✌🏻

k0rvusk0r4x commented:
The color is off: 
Branch: 
![image](https://user-images.githubusercontent.com/33426643/221260199-2dfc1a72-b471-48b6-b4f5-d002efdd034a.png)
Mockup: 
![image](https://user-images.githubusercontent.com/33426643/221260295-b4542a26-cdef-4b45-8d05-3966e4e95899.png)

Is this deliberate for accessibility or similar?

Mobile only:
Branch has the chosen filters wrapping: 
![image](https://user-images.githubusercontent.com/33426643/221263428-d21540c1-be9e-4e66-a7f6-865f415272ad.png)
On the mockup, they scroll: 
![image](https://user-images.githubusercontent.com/33426643/221263497-1ec1c484-ea5e-48c8-89bf-9604395b2e5d.png)

I think wrapping is better here, but it is a difference we want to be sure we are doing deliberately.

dev_block 😿 to assess these differences and decide if we want to change the speck or change the branch.

federicobadini commented:
> Is this deliberate for accessibility or similar?

It is for uniformity with the other chips that we have onto the website - we can update the mockup but that was discussed with @mmarcon91 

> I think wrapping is better here, but it is a difference we want to be sure we are doing deliberately.

Discussed [here](https://github.com/iFixit/react-commerce/pull/1375#discussion_r1112390541) :)

federicobadini commented:
un_dev_block 👍

masonmcelvain reviewed (approved):
CR ✌🏻 I'm uncertain why the tests are failing. I'm also noticing that builds on `main` are failing. I'm looking into it

masonmcelvain reviewed (dismissed):
CR ✌🏻

erinemay commented:
In general, it looks good. dev_block 🧷 on some questions and a couple minor things I noticed. 

<details><summary>1. Desktop: Show more/Show less not working</summary>

On a collection like https://react-commerce-git-refactor-product-list-facets-ifixit.vercel.app/Parts/iPhone, I see the item-type list is fully expanded, but show more/show less is still present and doesn't have any effect. I'm seeing it on Chrome, FF, and Safari.

https://user-images.githubusercontent.com/72166715/221687689-9a17226d-4a96-41e5-82d4-dc29436f981e.mp4

</details>


<details><summary>2. Desktop: Should the facet container resize dynamically?</summary>

This might have been discussed elsewhere, but the mockup appears to have the facet container dynamically sized (like the current implementation). This branch has it as the same static size, filling the vertical viewport space.

| Mockup/current | this branch |
| - | - |
| ![image](https://user-images.githubusercontent.com/72166715/221689695-c5e40204-46fb-4724-86e3-8299cae3ab59.png) | ![image](https://user-images.githubusercontent.com/72166715/221690204-ae228208-9fcc-4171-88c1-7cb42f894fec.png) |

</details>

<details><summary>5. Tablet: Facet container extends beyond the screen</summary>

This is the iPad 9th with Safari 15. It looks like the container keeps going until the end of the productlist.

![image](https://user-images.githubusercontent.com/72166715/221691174-28c1c348-55be-4538-9f72-9b2717446673.png)
 
</details>

<details><summary>4. Mobile: Should the filter list have some room to peek at the product list?</summary>

Or is this just a demonstration that it should open as a drawer?

| Mockup | this branch |
| - | - |
| ![image](https://user-images.githubusercontent.com/72166715/221693080-846e1730-fb27-4452-b706-6994168a0bf1.png) | ![image](https://user-images.githubusercontent.com/72166715/221693204-2c6d62b3-2e8f-44c7-9a72-768e034b2cf8.png) |




</details>

federicobadini commented:
> Desktop: Show more/Show less not working

Fixed

> Desktop: Should the facet container resize dynamically?

Fixed

> This is the iPad 9th with Safari 15. It looks like the container keeps going until the end of the productlist

Without the fix above, the facet container maxHeight was 100vh minus some margins. However on safari 100vh is not taking into account the space of the address bar, which reduces the available viewport.
So 100vh - when the bar is visible - is effectively more than the visible viewport height.
The fact that the container was sticky also created this "illusion" that its height was effectively equal to the product list height. Now that the fix above is in place we should have the same behaviour we currently have in production.

> Mobile: Should the filter list have some room to peek at the product list?

It would be nice, however it was excluded from this PR since we have several other drawers that would need to be styled too (cart, navigation etc). So it will be discussed in a separate PR

un_dev_block 👍

masonmcelvain reviewed (approved):
CR ✌🏻

masonmcelvain reviewed (dismissed):
CR ✌🏻

ardelato commented:
My bad, I accidentally pushed to this branch while trying to deflake the `parts_page_search` test 😓 . I undid the commit and had to force push.

ardelato commented:
CR 🚗 carry-over

erinemay commented:
The facets are working as expected and the above noted things are fixed. qa 🦖 

deploy_block 🧷 @federicobadini, I've got a weird one for you. 

I'm seeing this gap to the right side of some mobile device/browser combinations, but I'm not sure if it's an artifact of previewing and browserstack, or an actual bug. Are you seeing this? I haven't confirmed it's present on real devices. I'm asking around now for someone with an iPhone. I'm not seeing it on mobile Android + Chrome.

Here's an example problematic URL: https://react-commerce-git-refactor-product-list-facets-ifixit.vercel.app/Parts/Teenage_Engineering_OP-Z

And the main version: https://react-commerce.vercel.app/Parts/Teenage_Engineering_OP-Z

<details><summary>Screenshots and details</summary>
Browserstack - iPhone 14 + Safari 16

| this branch | main |
| - | - |
| ![image](https://user-images.githubusercontent.com/72166715/221993990-6e58835c-08f1-4763-849d-3b27d1fb378d.png) | ![image](https://user-images.githubusercontent.com/72166715/221997227-8621799d-c6bb-4ad4-9336-2726ce1a7235.png) |

FF dev tools site preview as iPhone 11 Pro (happening on MacOS and Windows)

| this branch | main |
| - | - |
| ![image](https://user-images.githubusercontent.com/72166715/221994357-7802b32f-ce5b-4aca-859c-ef2c81341577.png) | ![image](https://user-images.githubusercontent.com/72166715/221997554-2111f899-977d-41c7-923d-09862041b418.png) |



There isn't any obvious element that's expanding beyond the page limit. 

https://user-images.githubusercontent.com/72166715/221998127-06e01ed4-d8c6-46a3-8202-21be8f5eb05c.mp4

</details>

deltuh-vee commented:
I see it on an iPhone SE
![IMG_CF493CF1808D-1](https://user-images.githubusercontent.com/5289715/222001853-53ba6cb2-2929-4e30-bf42-f0941a5dae58.jpeg)

federicobadini commented:
> I've got a weird one for you.

The list and grid items were updated for the new layout in #1376 
Since it was merged, I am pretty confident that just updating from main will resolve that issue

federicobadini commented:
un_dev_block 👍

erinemay commented:
Awesome. Yep. That did it. 

QA 👍 

- On desktop, tablet, and mobile, the above issues have been resolved or addressed. 
- The filter facets look like the mockup and behave in an intuitive manner. 
- Adding selections works
- Removing selections (individually and all together) works
- The container responds dynamically
- Available facets refresh immediately when one is selected that changes availability of others
- In general, I poked around and added/removed/paginated, moved up and down the hierarchy and didn't find anything else of note.

deltuh-vee commented:
Carry Over QA 🎬

masonmcelvain reviewed (approved):
CR ✌🏻
